mplib/urdf_utils.py

```diff
@@ -0,0 +1,188 @@
+            if (cnt := (collision_cnt[i][j] + collision_cnt[j][i])) == num_samples:
+                if verbose:
+                    print(
+                        f"Ignore collision pair: ({link1}, {link2}), "
+                        "reason: Always collide"
+                    )
+                _ = ET.SubElement(
+                    root,
+                    "disable_collisions",
+                    attrib={"link1": link1, "link2": link2, "reason": "Always"},
+                )
+            elif cnt == 0:
+                if verbose:
+                    print(
+                        f"Ignore collision pair: ({link1}, {link2}), "
+                        "reason: Never collide"
+                    )
+                _ = ET.SubElement(
+                    root,
+                    "disable_collisions",
+                    attrib={"link1": link1, "link2": link2, "reason": "Never"},
+                )
+
+    return minidom.parseString(ET.tostring(root)).toprettyxml(indent="  ")
+
+
+def replace_urdf_package_keyword(
+    urdf_path: str | Path,
+    new_package_keyword: str = "",
+    *,
+    save_path: Optional[str | Path] = None,
+) -> Path:
+    """
+    Some ROS URDF files use package:// keyword to refer the package dir.
+    Replace it with the given string (default is empty)
+
+    :param urdf_path: Path to a Unified Robot Description Format file.
+    :param new_package_keyword: the string to replace 'package://' keyword
+    :param save_path: Path to save the modified URDF file.
+        If ``None``, create a temporary file.
+    :return: Path to the modified URDF file
+    """
+    if save_path is None:
+        _, save_path = tempfile.mkstemp(suffix=".urdf")
+    save_path = Path(save_path)
+    if not (save_dir := save_path.parent).is_dir():
+        save_dir.mkdir(parents=True)
+
+    with Path(urdf_path).open("r") as in_f:
+        if "package://" in (content := in_f.read()):
+            with save_path.open("w") as out_f:
+                out_f.write(content.replace("package://", new_package_keyword))
+    return save_path
+
+
+def generate_srdf(
+    urdf_path: str | Path,
+    new_package_keyword: str = "",
+    *,
+    num_samples=100000,
+    save_path: Optional[str | Path] = None,
+    verbose=False,
+) -> Path:
+    """
+    Generate SRDF from URDF similar to MoveIt2's setup assistant.
+
+    :param urdf_path: Path to a Unified Robot Description Format file.
+    :param new_package_keyword: the string to replace 'package://' keyword
+    :param num_samples: number of samples to find the link that will always collide
+    :param save_path: Path to save the generated SRDF file.
+        If ``None``, create a temporary file.
+    :param verbose: print debug info
+    :return: Path to the generated SRDF file
+    """
+    assert Path(urdf_path).is_file(), f"URDF file {urdf_path} does not exist"
+
+    # Replace 'package://' keyword
+    urdf_path = replace_urdf_package_keyword(urdf_path, new_package_keyword)
```

Comment:
This is a bit tricky. We probably want the srdf to be permanent so the user doesn't have to regenerate every single time. This will return a temporary path so next time the user wants to run the program again the srdf associated with this urdf will have the wrong name.